ddl/ddl_api.go

```diff
@@ -3920,25 +3919,34 @@ func checkAutoRandom(tableInfo *model.TableInfo, originCol *table.Column, specNe
 	switch {
 	case oldRandBits == newRandBits:
 		break
-	case oldRandBits == 0 || newRandBits == 0:
+	case newRandBits == 0:
 		return 0, ErrInvalidAutoRandom.GenWithStackByArgs(autoid.AutoRandomAlterErrMsg)
 	case autoid.MaxAutoRandomBits < newRandBits:
 		errMsg := fmt.Sprintf(autoid.AutoRandomOverflowErrMsg,
 			autoid.MaxAutoRandomBits, newRandBits, specNewColumn.Name.Name.O)
 		return 0, ErrInvalidAutoRandom.GenWithStackByArgs(errMsg)
+	case oldRandBits == 0:
+		isAutoIncPrimaryKey := mysql.HasAutoIncrementFlag(originCol.Flag) && originCol.IsPKHandleColumn(tableInfo)
+		if !isAutoIncPrimaryKey {
+			return 0, ErrInvalidAutoRandom.GenWithStackByArgs(autoid.AutoRandomAlterChangeFromAutoInc)
+		}
 	case oldRandBits < newRandBits:
 		break // Increasing auto_random shard bits is allowed.
 	case oldRandBits > newRandBits:
 		return 0, ErrInvalidAutoRandom.GenWithStackByArgs(autoid.AutoRandomDecreaseBitErrMsg)
 	}
 
-	if oldRandBits != 0 {
+	modifyingAutoRandCol := oldRandBits > 0 || newRandBits > 0
+	if modifyingAutoRandCol {
 		// Disallow changing the column field type.
 		if originCol.Tp != specNewColumn.Tp.Tp {
 			return 0, ErrInvalidAutoRandom.GenWithStackByArgs(autoid.AutoRandomModifyColTypeErrMsg)
 		}
-		// Disallow changing auto_increment on auto_random column.
-		if containsColumnOption(specNewColumn, ast.ColumnOptionAutoIncrement) != mysql.HasAutoIncrementFlag(originCol.Flag) {
+		if originCol.Tp != mysql.TypeLonglong {
+			return 0, ErrInvalidAutoRandom.GenWithStackByArgs(fmt.Sprintf(autoid.AutoRandomOnNonBigIntColumn, types.TypeStr(originCol.Tp)))
+		}
+		// Disallow changing from auto_random to auto_increment column.
+		if containsColumnOption(specNewColumn, ast.ColumnOptionAutoIncrement) && !mysql.HasAutoIncrementFlag(originCol.Flag) {
```

Comment:
Can this logic be removed? It has been confirmed in the getModifiableColumnJob method that it cannot support modifying the column from not_auto_increment to auto_increment.

Reply:
Even if adding auto_increment is supported, I hope the guy who implement it consider this problem carefully. So I think it is OK to check it again since there is no performance problem.